# Working on boston housing price dataset 


Regression problem to predict house price

In [7]:
from sklearn.datasets import load_boston

In [8]:
X, y = load_boston(return_X_y=True)

/Users/rajatchauhan/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the Californi

In [9]:
X

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]])

In [10]:
X.shape

(506, 13)

So we have total 13 features and number of records in the dataset are 506

In [11]:
y.shape

(506,)

# Using different regression algorithms on this training dataset

We will use these three algorithms:

1. Linear Regression
2. Decision Tree Regressor
3. Support Vector Regressor

In [17]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.svm import SVR

In [18]:
lr = LinearRegression()
dt = DecisionTreeRegressor()
svr = SVR()

Let us see individual performance of these algorithms:

In [20]:
estimators = [("Linear Regressor", lr) , ("Decision Trees", dt) , ("Support Vector Regressor", svr)]

For regression problems the scoring parameter we can use is R2 Score

In [22]:
from sklearn.model_selection import cross_val_score

In [27]:
import numpy as np

for estimator in estimators:
    scores = cross_val_score(estimator[1], X,y, cv = 10, scoring = 'r2')
    print(estimator[0], "score ",np.round(scores.mean(),2))

Linear Regressor score  0.2
Decision Trees score  -0.15
Support Vector Regressor score  -0.41


So the models are performing very bad, but lets see how does the voting regressor performs

# Voting regressor

In [26]:
from sklearn.ensemble import VotingRegressor

In [30]:
vr = VotingRegressor(estimators)
scores = cross_val_score(vr, X,y, cv = 10, scoring = 'r2')
print("Voting Regressor score", np.round(scores.mean(),2))

Voting Regressor score 0.46


The improvement is good

# Experimenting with weights

In [34]:
for i in range(1,4):
    for j in range(1,4):
        for k in range(1,4):
            vr = VotingRegressor(estimators, weights = [i,j,k])
            scores = cross_val_score(vr, X,y, cv = 10, scoring = 'r2')
            print("For i = {}, j = {}, k = {} Score:".format(i,j,k), np.round(scores.mean(),2))

For i = 1, j = 1, k = 1 Score: 0.45
For i = 1, j = 1, k = 2 Score: 0.35
For i = 1, j = 1, k = 3 Score: 0.26
For i = 1, j = 2, k = 1 Score: 0.39
For i = 1, j = 2, k = 2 Score: 0.39
For i = 1, j = 2, k = 3 Score: 0.36
For i = 1, j = 3, k = 1 Score: 0.36
For i = 1, j = 3, k = 2 Score: 0.37
For i = 1, j = 3, k = 3 Score: 0.37
For i = 2, j = 1, k = 1 Score: 0.45
For i = 2, j = 1, k = 2 Score: 0.41
For i = 2, j = 1, k = 3 Score: 0.36
For i = 2, j = 2, k = 1 Score: 0.42
For i = 2, j = 2, k = 2 Score: 0.44
For i = 2, j = 2, k = 3 Score: 0.41
For i = 2, j = 3, k = 1 Score: 0.41
For i = 2, j = 3, k = 2 Score: 0.44
For i = 2, j = 3, k = 3 Score: 0.36
For i = 3, j = 1, k = 1 Score: 0.44
For i = 3, j = 1, k = 2 Score: 0.4
For i = 3, j = 1, k = 3 Score: 0.39
For i = 3, j = 2, k = 1 Score: 0.47
For i = 3, j = 2, k = 2 Score: 0.47
For i = 3, j = 2, k = 3 Score: 0.42
For i = 3, j = 3, k = 1 Score: 0.42
For i = 3, j = 3, k = 2 Score: 0.45
For i = 3, j = 3, k = 3 Score: 0.45


# Using same algorithm instances

Let us use decision trees with different depths

In [35]:
dt1 = DecisionTreeRegressor(max_depth=1)
dt2 = DecisionTreeRegressor(max_depth=3)
dt3 = DecisionTreeRegressor(max_depth=5)
dt4 = DecisionTreeRegressor(max_depth=7)
dt5 = DecisionTreeRegressor(max_depth=None)

In [37]:
estimators = [("Decision Tree 1", dt1), ("Decision Tree 2", dt2), ("Decision Tree 3", dt3), 
              ("Decision Tree 4", dt4), ("Decision Tree 5", dt5)]

Let us see individual performance of these decision trees first

In [38]:
for estimator in estimators:
    scores = cross_val_score(estimator[1], X,y, cv = 10, scoring='r2')
    print(estimator[0], np.round(scores.mean(),2))

Decision Tree 1 -0.85
Decision Tree 2 -0.11
Decision Tree 3 0.03
Decision Tree 4 0.08
Decision Tree 5 -0.07


Again not so good result

Let us use Voting regressor on top of these decision trees now

In [39]:
vr = VotingRegressor(estimators)
scores = cross_val_score(vr, X,y, cv = 10, scoring='r2')
print("Voting regressor R2 score: ", np.round(scores.mean(),2))

Voting regressor R2 score:  0.21


So, great thing the scores are increasing